File: optimizer.js
// Prompt Optimizer - JavaScript/WebAssembly version for browser

// Common English stopwords
const STOPWORDS = new Set([
    'a', 'about', 'above', 'after', 'again', 'against', 'all', 'am', 'an', 'and', 'any', 'are',
    'as', 'at', 'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but',
    'by', 'can', 'cannot', 'could', 'did', 'do', 'does', 'doing', 'down', 'during', 'each',
    'few', 'for', 'from', 'further', 'had', 'has', 'have', 'having', 'he', 'her', 'here',
    'hers', 'herself', 'him', 'himself', 'his', 'how', 'i', 'if', 'in', 'into', 'is', 'it',
    'its', 'itself', 'just', 'me', 'might', 'more', 'most', 'myself', 'my', 'no', 'nor', 'not',
    'of', 'off', 'on', 'once', 'only', 'or', 'other', 'our', 'ours', 'ourselves', 'out', 'over',
    'own', 'same', 'she', 'should', 'so', 'some', 'such', 'than', 'that', 'the', 'their',
    'theirs', 'them', 'themselves', 'then', 'there', 'these', 'they', 'this', 'those', 'to',
    'too', 'under', 'until', 'up', 'we', 'were', 'what', 'when', 'where', 'which', 'while',
    'who', 'whom', 'why', 'with', 'would', 'you', 'your', 'yours', 'yourself', 'yourselves'
]);

// Approximate token counter (based on OpenAI's tiktoken heuristic)
function countTokens(text) {
    // Rough approximation: 1 token ≈ 4 characters on average
    // This is a heuristic; actual tiktoken may differ
    const words = text.trim().split(/\s+/);
    let tokenCount = 0;
    
    words.forEach(word => {
        // Heuristic: average word is ~0.75 tokens, with minimum of 1
        tokenCount += Math.max(1, Math.ceil(word.length / 4));
    });
    
    return tokenCount;
}

// TF-IDF based keyword extraction
function extractKeywords(text, topN = 10) {
    const words = text.toLowerCase().split(/\s+/);
    const uniqueWords = [...new Set(words)];
    
    // Calculate term frequency
    const tf = {};
    uniqueWords.forEach(word => {
        tf[word] = (tf[word] || 0) + 1;
    });
    
    // Calculate IDF (inverse document frequency) - simplified version
    const totalWords = uniqueWords.length;
    const idf = {};
    uniqueWords.forEach(word => {
        // Simple heuristic: less common words get higher IDF
        const frequency = tf[word] / uniqueWords.length;
        idf[word] = Math.log(1 / (frequency + 0.0001));
    });
    
    // Calculate TF-IDF score
    const scores = uniqueWords.map(word => ({
        word: word,
        score: tf[word] * idf[word]
    }));
    
    // Sort by score and return top N
    return scores
        .sort((a, b) => b.score - a.score)
        .slice(0, topN)
        .map(item => [item.word, item.score]);
}

// Model pricing configuration (per 1k tokens)
const MODELS = {
    "gemini-1.5-flash": {
        name: "Google Gemini 1.5 Flash",
        input_cost_per_1k: 0.000125,
    },
    "gemini-1.5-pro": {
        name: "Google Gemini 1.5 Pro",
        input_cost_per_1k: 0.00125,
    },
    "gemini-2.0": {
        name: "Google Gemini 2.0",
        input_cost_per_1k: 0.002,
    },
    "gpt-4o-mini": {
        name: "OpenAI GPT-4o Mini",
        input_cost_per_1k: 0.00015,
    },
    "gpt-4o": {
        name: "OpenAI GPT-4o",
        input_cost_per_1k: 0.005,
    },
    "gpt-5.3": {
        name: "OpenAI GPT-5.3",
        input_cost_per_1k: 0.05,
    },
    "gpt-5.4": {
        name: "OpenAI GPT-5.4",
        input_cost_per_1k: 0.06,
    },
    "gpt-4": {
        name: "OpenAI GPT-4",
        input_cost_per_1k: 0.03,
    },
    "claude-3.5-haiku": {
        name: "Anthropic Claude 3.5 Haiku",
        input_cost_per_1k: 0.00080,
    },
    "claude-3.5-sonnet": {
        name: "Anthropic Claude 3.5 Sonnet",
        input_cost_per_1k: 0.003,
    },
    "claude-3.5-opus": {
        name: "Anthropic Claude 3.5 Opus",
        input_cost_per_1k: 0.015,
    },
    "claude-opus-4.6": {
        name: "Anthropic Claude Opus 4.6",
        input_cost_per_1k: 0.02,
    },
    "mistral-large": {
        name: "Mistral Large",
        input_cost_per_1k: 0.008,
    },
    "llama-2-7b": {
        name: "Open Source - Llama 2 7B",
        input_cost_per_1k: 0.0,
    },
};

// Main optimization function
function optimizePrompt(prompt, targetTokens = 50, alpha = 0.1, model = "gemini-1.5-flash") {
    // 1. Clean text - remove stopwords and punctuation
    const cleanedWords = prompt
        .toLowerCase()
        .split(/\s+/)
        .filter(word => {
            // Remove if stopword or pure punctuation
            const isStopword = STOPWORDS.has(word.replace(/[^\w]/g, ''));
            const isPunctuation = /^[^\w]+$/.test(word);
            return !isStopword && !isPunctuation;
        });
    
    const cleanedText = cleanedWords.join(' ');
    
    if (cleanedWords.length === 0) {
        return {
            optimized_prompt: "",
            metrics: {
                input_tokens: 0,
                final_tokens: 0,
                target_threshold: targetTokens,
                alpha: alpha,
                model: model,
                model_name: MODELS[model]?.name || "Unknown",
                cents_saved: 0,
                reduction_percent: 0,
                is_estimate: true,
            }
        };
    }
    
    // 2. Extract keywords using TF-IDF
    const keywords = extractKeywords(cleanedText, cleanedWords.length);
    const saliencyMap = {};
    keywords.forEach(([word, score]) => {
        saliencyMap[word] = score;
    });
    
    // 3. Unconstrained objective optimization
    const selected = [];
    let currentTokens = 0;
    const wordBonus = 0.05;
    
    // Get original words (case-preserved)
    const originalWords = prompt.split(/\s+/);
    
    // Create candidates with metadata
    const candidates = [];
    originalWords.forEach(word => {
        const cleanWord = word.toLowerCase().replace(/[^\w]/g, '');
        const wordTokens = countTokens(word);
        const saliency = saliencyMap[cleanWord] || 0;
        
        candidates.push({
            word: word,
            cleanWord: cleanWord,
            tokens: wordTokens,
            saliency: saliency,
            valueDensity: saliency / Math.max(wordTokens, 1)
        });
    });
    
    // Sort by value density
    candidates.sort((a, b) => b.valueDensity - a.valueDensity);
    
    // Select words using exponential penalty
    candidates.forEach(candidate => {
        const penalty = Math.exp(alpha * (currentTokens + candidate.tokens - targetTokens));
        const utility = (candidate.saliency + wordBonus) - penalty;
        
        if (utility > 0) {
            selected.push(candidate);
            currentTokens += candidate.tokens;
        }
    });
    
    // Sort selected words back to original order
    const selectedSet = new Set(selected.map((_, i) => originalWords.indexOf(selected[i].word)));
    const optimizedWords = originalWords.filter((_, i) => {
        for (let s of selected) {
            if (s.word === originalWords[i]) {
                selected.shift();
                return true;
            }
        }
        return false;
    });
    
    // Reconstruct maintaining original order
    const selectedIndices = [];
    selected.forEach(s => {
        for (let i = 0; i < originalWords.length; i++) {
            if (originalWords[i] === s.word && !selectedIndices.includes(i)) {
                selectedIndices.push(i);
                break;
            }
        }
    });
    selectedIndices.sort((a, b) => a - b);
    const optimizedPrompt = selectedIndices.map(i => originalWords[i]).join(' ');
    
    // 4. Calculate savings
    const inputTokens = countTokens(prompt);
    const savedTokens = inputTokens - currentTokens;
    
    const modelInfo = MODELS[model] || MODELS["gemini-1.5-flash"];
    const costPer1k = modelInfo.input_cost_per_1k;
    
    const originalCost = (inputTokens / 1000) * costPer1k * 100; // in cents
    const optimizedCost = (currentTokens / 1000) * costPer1k * 100; // in cents
    const centsSaved = originalCost - optimizedCost;
    
    const reductionPercent = inputTokens > 0 ? Math.round((savedTokens / inputTokens * 100) * 10) / 10 : 0;
    
    return {
        optimized_prompt: optimizedPrompt,
        metrics: {
            input_tokens: inputTokens,
            final_tokens: currentTokens,
            target_threshold: targetTokens,
            alpha: alpha,
            model: model,
            model_name: modelInfo.name,
            cents_saved: Math.round(centsSaved * 10000) / 10000,
            reduction_percent: reductionPercent,
            is_estimate: true,
        }
    };
}

// Export for use in browser or Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { optimizePrompt, countTokens, extractKeywords, MODELS };
}

File: optimizer.py
import math
from pydantic import BaseModel
import tiktoken
from keybert import KeyBERT
import spacy
import json
import sys
import os

# 1. Silence TensorFlow/HuggingFace logging
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
os.environ['TOKENIZERS_PARALLELISM'] = 'false'

import warnings
warnings.filterwarnings("ignore")

# 2. Redirect ALL library startup 'print' statements to stderr
# so they don't corrupt the final output
class SuppressStdout:
    def __enter__(self):
        self._original_stdout = sys.stdout
        sys.stdout = sys.stderr

    def __exit__(self, exc_type, exc_val, exc_tb):
        sys.stdout = self._original_stdout

with SuppressStdout():
    nlp = spacy.load("en_core_web_sm", disable=["ner", "parser", "lemmatizer"])
    kw_model = KeyBERT()
    tokenizer = tiktoken.get_encoding("cl100k_base")

# Model pricing configuration (per 1k tokens)
# These are estimates - actual costs may vary based on rate changes
MODELS = {
    "gemini-1.5-flash": {
        "name": "Google Gemini 1.5 Flash",
        "input_cost_per_1k": 0.000125,
    },
    "gpt-4o-mini": {
        "name": "OpenAI GPT-4o Mini",
        "input_cost_per_1k": 0.00015,
    },
    "gpt-4": {
        "name": "OpenAI GPT-4",
        "input_cost_per_1k": 0.03,
    },
    "claude-3.5-haiku": {
        "name": "Anthropic Claude 3.5 Haiku",
        "input_cost_per_1k": 0.00080,
    },
    "claude-3.5-sonnet": {
        "name": "Anthropic Claude 3.5 Sonnet",
        "input_cost_per_1k": 0.003,
    },
    "llama-2-7b": {
        "name": "Open Source - Llama 2 7B",
        "input_cost_per_1k": 0.0,
    },
}

class PromptRequest(BaseModel):
    prompt: str
    target_tokens: int = 50
    alpha: float = 0.1
    model: str = "gemini-1.5-flash" 

def optimize_unconstrained(req: PromptRequest):
    doc = nlp(req.prompt)

    # 1. Extraction & Scoring
    tokens = [t for t in doc if (not t.is_stop or t.dep_ == "neg") and not t.is_punct and t.pos_ != "INTJ"]
    text_cleaned = " ".join([t.text for t in tokens])

    if not tokens:
        return {"optimized_prompt": "", "metrics": {"input_tokens": 0, "final_tokens": 0}}

    keywords = kw_model.extract_keywords(text_cleaned, top_n=len(tokens))
    saliency_map = {kw[0].lower(): kw[1] for kw in keywords}

    # 2. Unconstrained Objective Optimization
    selected = []
    current_tokens = 0
    word_bonus = 0.05 # Encourages maximizing word count

    # Sort by Value Density
    candidates = sorted(tokens, key=lambda t: saliency_map.get(t.text.lower(), 0) / max(len(tokenizer.encode(t.text)), 1), reverse=True)

    for t in candidates:
        w_len = len(tokenizer.encode(t.text))
        saliency = saliency_map.get(t.text.lower(), 0)

        # Exponential Penalty (Normalizing to the same scale as Saliency)
        # Saliency is 0-1, so we want penalty to start small and cross 1.0 at target
        penalty = math.exp(req.alpha * (current_tokens + w_len - req.target_tokens))

        # Normalized Objective
        utility = (saliency + word_bonus) - penalty

        if utility > 0:
            selected.append(t)
            current_tokens += w_len

    selected.sort(key=lambda t: t.i)
    optimized_prompt = " ".join([t.text for t in selected])

    # 3. Savings Calculation
    input_tokens = len(tokenizer.encode(req.prompt))
    saved_tokens = input_tokens - current_tokens

    # Get model pricing
    model_info = MODELS.get(req.model, MODELS["gemini-1.5-flash"])
    cost_per_1k = model_info["input_cost_per_1k"]

    original_cost = (input_tokens / 1000) * cost_per_1k * 100  # in cents
    optimized_cost = (current_tokens / 1000) * cost_per_1k * 100  # in cents
    cents_saved = original_cost - optimized_cost

    reduction_percent = round((saved_tokens / input_tokens * 100), 1) if input_tokens > 0 else 0

    return {
        "optimized_prompt": optimized_prompt,
        "metrics": {
            "input_tokens": input_tokens,
            "final_tokens": current_tokens,
            "target_threshold": req.target_tokens,
            "alpha": req.alpha,
            "model": req.model,
            "model_name": model_info["name"],
            "cents_saved": round(cents_saved, 4),
            "reduction_percent": reduction_percent,
            "is_estimate": True,
        }
    }

if __name__ == "__main__":
    try:
        raw_input = sys.argv[1]
        input_data = json.loads(raw_input)

        # FIX: Ensure we extract the string from the dict if needed
        if isinstance(input_data, dict):
            prompt_text = input_data.get("prompt", "")
            target_tokens = input_data.get("target_tokens", 50)
            alpha = input_data.get("alpha", 0.1)
            model = input_data.get("model", "gemini-1.5-flash")
        else:
            prompt_text = input_data
            target_tokens = 50
            alpha = 0.1
            model = "gemini-1.5-flash"

        req = PromptRequest(prompt=prompt_text, target_tokens=target_tokens, alpha=alpha, model=model)
        result = optimize_unconstrained(req)

        # Use sys.stdout.write to avoid extra newlines
        sys.stdout.write(json.dumps(result))

    except Exception as e:
        sys.stderr.write(f"Runtime Error: {str(e)}")
        sys.exit(1)
        